[CLI] Check with multiple arguments

jaudiger opened the pull request:
Part of https://github.com/brioche-dev/brioche/issues/75

I added the ability to set more than one registry and project when calling `brioche check` command. This is separated in two commits. The first updates the check command, while the second commit refactor the code of the format command to mimic what was done with the check.

This has been done to extend more easily the next commands, cited in https://github.com/brioche-dev/brioche/issues/75.

jaudiger commented:
Conflict with https://github.com/brioche-dev/brioche/pull/90. Depending on the one merged first, I'll update the code of the second PR.

kylewlacy commented on crates/brioche/src/check.rs:
```diff
@@ -1,51 +1,109 @@
+
+    match result {
+        Ok(()) => {
+            println!("No errors found on {project_name} 🎉");
```
Nitpick: rewording

```suggestion
            println!("No errors found in {project_name} 🎉");
```

jaudiger replied:
Thanks, I fixed the typo

kylewlacy commented on crates/brioche/src/check.rs:
```diff
@@ -1,51 +1,109 @@
+    }
 
-        let result = checked.ensure_ok(brioche_core::script::check::DiagnosticLevel::Message);
+    // Loop over the registries
```
Nitpick: rewording. `-r` takes a registry project (that is, a "project that should get downloaded from the registry") not a registry itself (and the flag is named `registry` to keep it shorter... although renaming the struct field while keeping the flag the same might be a good idea)

```suggestion
    // Loop over the registry projects
```

jaudiger replied:
I updated the registry argument to registry_project, and I added a clap identifier to not change the CLI interface.

kylewlacy reviewed (commented):
LGTM, just had two minor suggestions for some things to reword slightly

kylewlacy commented on crates/brioche/src/check.rs:
```diff
@@ -1,51 +1,114 @@
+use std::path::PathBuf;
 use std::process::ExitCode;
+use std::vec;
```
```suggestion
```

kylewlacy reviewed (approved):
LGTM (I just noticed one unnecessary import but will merge after it gets removed, plus the remaining conflicts)

jaudiger commented:
@kylewlacy do you prefer a rebase or a merge to resolve the merge conflicts ?

kylewlacy commented:
> @kylewlacy do you prefer a rebase or a merge to resolve the merge conflicts ?

Generally, I prefer merges specifically because GitHub's PR reviews work better with it (I can click "show changes since last review" and get a nice diff, but that tends to break when force pushing / rebasing). I always use "Squash and merge" when merging PRs anyway, so the final result in the `main` branch ends up basically the same either way

jaudiger commented:
> > @kylewlacy do you prefer a rebase or a merge to resolve the merge conflicts ?
> 
> Generally, I prefer merges specifically because GitHub's PR reviews work better with it (I can click "show changes since last review" and get a nice diff, but that tends to break when force pushing / rebasing). I always use "Squash and merge" when merging PRs anyway, so the final result in the `main` branch ends up basically the same either way

Sorry..... I didn't see when I was preparing the change locally... I'm used to rebase, since I like having work commit. But I'll now switch to the merge strategy for my other PRs! Basically, what I did was just to integrate the reporter stuff inside the three commits. If for the sake of review it doesn't suit you, please let me know.

At least, for this time, the compare option seems to kind of work: https://github.com/brioche-dev/brioche/compare/e12668f5dc67886b70e1120773ea087deb9de9af..9a7244a230fcf59bb6402b04bc26778f5e7218cc (except I didn't touch the publish file)

jaudiger commented on crates/brioche/src/check.rs:
```diff
@@ -1,51 +1,128 @@
+) -> Result<bool, anyhow::Error> {
+    let num_lockfiles_updated = projects.commit_dirty_lockfiles().await?;
+    if num_lockfiles_updated > 0 {
+        tracing::info!(num_lockfiles_updated, "updated lockfiles");
```
One thing, I didn't touch this line, but should I?

kylewlacy replied:
I think this is fine as-is, is there a reason it should be changed?

jaudiger replied:
No, I don't think so. I just wanted to raise the attention to this single call to `tracing::info`.

kylewlacy commented:
> Sorry..... I didn't see when I was preparing the change locally... I'm used to rebase, since I like having work commit. But I'll now switch to the merge strategy for my other PRs! Basically, what I did was just to integrate the reporter stuff inside the three commits. If for the sake of review it doesn't suit you, please let me know.

Not a problem 😄 It's basically just a personal preference, but more important is making sure it's a project where you and other contributors can use a workflow you're comfortable with first and foremost (which is also why I always go with "Squash and merge" when merging PRs into main, it seems like the middle ground between folks that like "merge" vs. "rebase", keeping the history clean while still including the commit messages and a link to the PR)
